### ISSUE TITLE
Referencing equations in separate files

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1210]
I am using Sphinx to write the documentation for a math solver library, and often need to reference mathematical equations in subsequent portions of the text.  While this is easy with the :eq:`equation_label` syntax, it only works for equations in the same file.  I realize that there are "workarounds" by setting a fixed reference label preceding said equation, but this is a very poor substitute for just having a global equation numbering system and global equation cross-referencing.

A number of discussion forums have suggested that someone submit "a bug/feature request to get this changed," but I don't know if this has happened, and a quick search with the issue tracker came up with nothing.

I don't imagine that this should be too difficult to implement, and I'm very much surprised that it isn't the default behavior already (since that's how it works in LaTeX).

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1210
- Originally reported by: Daniel Reynolds
- Originally created at: 2013-07-11T02:43:01.365

**Comment by shimizukawa** [comment]
_From Andrea Cassioli on 2014-07-07 13:34:11+00:00_

Hi,
I guess I have the same problem as Daniel. Is there any update on that? This is really a missing feature that we would like to see asap!

**Comment by shimizukawa** [comment]
_From Andrea Cassioli on 2014-08-19 12:40:43+00:00_

Update:

One of the issue is that equation labels are of the form docname:equationlabel. Dropping the docname solves the issue at least for the Latex builder.

**Comment by andreacassioli** [comment]
Hi,
is there any update on this issue? 

In the html writer the problem is that the code does

```
def html_visit_eqref(self, node):
    self.body.append('<a href="#equation-%s">' % node['target'])
```

and thus it always generates local references to the page. Could we imagine to store a dictionary of labels in the application main object?

**Comment by andreacassioli** [comment]
OK, 
I have made an ack to mathbase that seems to work: I assume my equation to be referenced either as `:eq:`label`` and in this case is the local doc; or as `:eq:`filename:label`` and then I can make a correct link. For HTML the code looks like

```
def html_visit_eqref(self, node):

    node_target= node['target'].encode('ascii','replace')
    current_doc=  node['docname'].encode('ascii','replace')

    if ':' in node_target:
        doc_target, link_target = node_target.split(':')
    else:
        doc_target= current_doc
        link_target= node_target

    if self.builder.name == 'json':
        link= ( '%s#%s' % (self.builder.get_relative_uri(current_doc, doc_target), link_target) )
        self.body.append('<a href="'+link+'">')
    else:
        self.body.append('<a href="%s.html#%s">' % ( doc_target, link_target))

```

Does it make sense? In case, how should I do a pull request?

**Comment by andreacassioli** [comment]
Hi,
after the release of Sphinx 1.3 I think that a very nice way to solve this issue is to use for the equations the same machinery of :numref: The role :eq: could be just an alias for :numref: and this will ensure back compatibility. Would it be a good solutions?

**Comment by xuhdev** [comment]
@andreacassioli Yes, I agree that if equations can just behave like figures it would be nice.

**Comment by andreacassioli** [comment]
Just stumbled again against this issue....

What would take to extend numref to equations? 

I guess math directive should implement `:name:` option and register the link. Then extend the writer to generate the right link label.

**Comment by andreacassioli** [comment]
Hi,
I am kind of thinking that maybe we just need to store the full link of the equation. Indeed, the generated link is just local, with no file information. What about change this line

https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/mathbase.py#L103

to include the full path? Would it be a quick and dirty hack?

**Comment by tk0miya** [comment]
I merged #2861 into master branch now. It will be fixed at next major release (v1.5).
Thanks,

**Comment by danrings** [comment]
Is there a toggle to switch this "enhancement" off? It creates quite a lot of extra work to change all existing labels which used to be unique inside a single file but are now supposed to be unique across all files. I think there is a stronger use case for NOT referencing across files by default.
